contracts/schemes/SchemeFactory.sol

```diff
@@ -0,0 +1,173 @@
+
+import "@daostack/infra-experimental/contracts/votingMachines/IntVoteInterface.sol";
+import "@daostack/infra-experimental/contracts/votingMachines/VotingMachineCallbacksInterface.sol";
+import "../votingMachines/VotingMachineCallbacks.sol";
+import "../utils/DAOFactory.sol";
+import "@openzeppelin/upgrades/contracts/Initializable.sol";
+
+
+/**
+ * @title A factory and registrar for Schemes for organizations
+ * @dev The SchemeFactory is used for deploying and registering schemes to organisations
+ */
+contract SchemeFactory is Initializable, VotingMachineCallbacks, ProposalExecuteInterface {
+    event NewSchemeProposal(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        address indexed _intVoteInterface,
+        string _schemeName,
+        bytes _schemeData,
+        uint64[3] _packageVersion,
+        bytes4 _permissions,
+        address _schemeToReplace,
+        string _descriptionHash
+    );
+
+    event ProposalExecuted(address indexed _avatar, bytes32 indexed _proposalId, int256 _param);
+    event ProposalDeleted(address indexed _avatar, bytes32 indexed _proposalId);
+
+    // a proposal to add or remove a scheme to/from the an organization
+    struct Proposal {
+        string schemeName;
+        bytes schemeData;
+        uint64[3] packageVersion;
+        address schemeToReplace;
+        bytes4 permissions;
+    }
+
+    mapping(bytes32=>Proposal) public proposals;
+
+    IntVoteInterface public votingMachine;
+    bytes32 public voteParams;
+    Avatar public avatar;
+    DAOFactory public daoFactory;
+
+    /**
+     * @dev initialize
+     * @param _avatar the avatar this scheme referring to.
+     * @param _votingMachine the voting machines address to
+     * @param _voteParams voting machine parameters to register scheme.
+     */
+    function initialize(
+        Avatar _avatar,
+        IntVoteInterface _votingMachine,
+        bytes32 _voteParams,
+        DAOFactory _daoFactory
+    )
+    external
+    initializer
+    {
+        require(_avatar != Avatar(0), "avatar cannot be zero");
+        avatar = _avatar;
+        votingMachine = _votingMachine;
+        voteParams = _voteParams;
+        daoFactory = _daoFactory;
+    }
+
+    /**
+    * @dev execution of proposals, can only be called by the voting machine in which the vote is held.
+    * @param _proposalId the ID of the voting in the voting machine
+    * @param _decision the voting result, 1 yes and 2 is no.
+    * @return true (if function did not revert)
+    */
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    returns(bool) {
+        Proposal memory proposal = proposals[_proposalId];
+        delete proposals[_proposalId];
+        emit ProposalDeleted(address(avatar), _proposalId);
+        Controller controller = Controller(avatar.owner());
+        if (_decision == 1) {
+            if (bytes(proposal.schemeName).length > 0) {
```

Comment:
does proposal.schemeName != "" does not work ?

Reply:
no